pkg/audit/stats_reporter.go

```diff
@@ -2,111 +2,102 @@ package audit
+	auditDurationM    metric.Float64Histogram
+	lastRunStartTimeM metric.Float64ObservableGauge
+	lastRunEndTimeM   metric.Float64ObservableGauge
+	endTime           time.Time
+	latency           time.Duration
+	startTime         time.Time
+	meter             metric.Meter
 )
 
 func init() {
-	if err := register(); err != nil {
+	var err error
+	meter = otel.GetMeterProvider().Meter("gatekeeper")
+
+	violationsM, err = meter.Int64ObservableGauge(
+		violationsMetricName,
+		metric.WithDescription("Total number of audited violations"),
+	)
+
+	if err != nil {
 		panic(err)
 	}
-}
 
-func register() error {
-	views := []*view.View{
-		{
-			Name:        violationsMetricName,
-			Measure:     violationsM,
-			Aggregation: view.LastValue(),
-			TagKeys:     []tag.Key{enforcementActionKey},
-		},
-		{
-			Name:        auditDurationMetricName,
-			Measure:     auditDurationM,
-			Aggregation: view.Distribution(1*60, 3*60, 5*60, 10*60, 15*60, 20*60, 40*60, 80*60, 160*60, 320*60),
-		},
-		{
-			Name:        lastRunStartTimeMetricName,
-			Measure:     lastRunStartTimeM,
-			Aggregation: view.LastValue(),
-		},
-		{
-			Name:        lastRunEndTimeMetricName,
-			Measure:     lastRunEndTimeM,
-			Aggregation: view.LastValue(),
-		},
+	auditDurationM, err = meter.Float64Histogram(
+		auditDurationMetricName,
+		metric.WithDescription("Latency of audit operation in seconds"))
+	if err != nil {
+		panic(err)
 	}
-	return view.Register(views...)
-}
 
-func (r *reporter) reportTotalViolations(enforcementAction util.EnforcementAction, v int64) error {
-	ctx, err := tag.New(
-		context.Background(),
-		tag.Insert(enforcementActionKey, string(enforcementAction)))
+	lastRunStartTimeM, err = meter.Float64ObservableGauge(
+		lastRunStartTimeMetricName,
+		metric.WithDescription("Timestamp of last audit run starting time"),
+	)
 	if err != nil {
-		return err
+		panic(err)
 	}
 
-	return r.report(ctx, violationsM.M(v))
-}
-
-func (r *reporter) reportLatency(d time.Duration) error {
-	ctx, err := tag.New(context.Background())
+	lastRunEndTimeM, err = meter.Float64ObservableGauge(
+		lastRunEndTimeMetricName,
+		metric.WithDescription("Timestamp of last audit run ending time"),
+	)
 	if err != nil {
-		return err
+		panic(err)
 	}
+}
 
-	return r.report(ctx, auditDurationM.M(d.Seconds()))
+func (r *reporter) registerCallback() error {
+	_, err1 := meter.RegisterCallback(r.reportTotalViolations, violationsM)
+	_, err2 := meter.RegisterCallback(r.reportRunEnd, lastRunEndTimeM)
+	_, err3 := meter.RegisterCallback(r.reportRunStart, lastRunStartTimeM)
+	return errors.Join(err1, err2, err3)
 }
 
-func (r *reporter) reportRunStart(t time.Time) error {
-	ctx, err := tag.New(context.Background())
-	if err != nil {
-		return err
+func (r *reporter) reportTotalViolations(_ context.Context, o metric.Observer) error {
+	for k, v := range totalViolationsPerEnforcementAction {
+		o.ObserveInt64(violationsM, v, metric.WithAttributes(attribute.String(enforcementActionKey, string(k))))
```

Comment:
To reduce the delta and to ensure correctness, we should keep all state inside the `reporter` object, with `reportTotalViolations()` and similar used to alter that internal state.

In general, "reportX"-type functions should still be called by the code being observed.

We should have a different mechanism for the callbacks.

Reply:
Can you elaborate more?

What do you mean by different mechanisms for the callbacks?

Reply:
Sorry, that was vague. The general issue is the use of global mutable state across threads.

reportX-type functions (e.g. reportTotalViolations()) should still be called by the system being observed (in this case audit manager).

Right now, you are using reportX-type functions both as callbacks (as in this instance), and in the original fashion (e.g. `reportLatency` still behaves the same).

We apparently still need callbacks (AFAICT otel requires them), but they should not be `reportX` functions. Either anonymous functions or something like `observeX` might be good. Having a predictable naming and structure will make it easier for others to grok the code later.

IMO otel should have had a consistent API for their metrics. They didn't do that, but fortunately we have the `reporter` abstraction to paper over this.